dedupe/datamodel.py

```diff
@@ -183,30 +178,35 @@ def typify_variables(
         try:
             variable_class = VARIABLE_CLASSES[variable_type]
         except KeyError:
+            valid = ", ".join(VARIABLE_CLASSES)
             raise KeyError(
-                "Field type %s not valid. Valid types include %s"
-                % (definition["type"], ", ".join(VARIABLE_CLASSES))
+                f"Variable type {variable_type} not valid. Valid types include {valid}"
             )
-
         variable_object = variable_class(definition)
-        assert isinstance(variable_object, FieldVariable)
-
-        primary_variables.append(variable_object)
+        assert isinstance(variable_object, Variable)
+        variables.append(variable_object)
 
-        if hasattr(variable_object, "higher_vars"):
-            all_variables.extend(variable_object.higher_vars)
-        else:
-            variable_object = cast(Variable, variable_object)
-            all_variables.append(variable_object)
-
-    if only_custom:
+    no_blocking_variables = all(
```

Comment:
Continuing from https://github.com/dedupeio/dedupe/pull/1098#discussion_r975730820

I think it would be better if this was more duck-typed though, such as 
`no_blocking_variables = not any(len(v.predicates) for v in variables)`.

That would be a slight change though, since `Variable.predicates` is filled with an `ExistsPredicate` for any var def with the `has missing` flag. Perhaps this is OK? But I don't think so.

I think really there needs to be a more official API where variables declare the predicates they export.

Reply:
the code that add the ExistsPredicate should only really be on FieldPredicates as it depends on there being a field.

https://github.com/dedupeio/dedupe/blob/main/dedupe/variables/base.py#L37

if we move that code, then we could do pretty much what you are saying.

```python
no_blocking_variables = not any(len(v.predicates) for v in variables)
```

we might also want to have a check like

```python
only_exists = all(isinstance(pred, ExistsPredicate) for v in variables for pred in v.predicates)
if only_exists:
    warning.warn('You are only using the Exists Predicate which is usually pretty bad')
```